src/cc/bcc_elf.c

```diff
@@ -196,20 +201,254 @@ static int listsymbols(Elf *e, bcc_elf_symcb callback, void *payload) {
+  if (!buf) {
+    close(fd);
+    return 0;
+  }
+
+  actual = gnu_debuglink_crc32(0, buf, st.st_size);
+
+  munmap(buf, st.st_size);
+  close(fd);
+  return actual == crc;
+}
+
+/*
+  @name       the name of the debug information file
+  @binpath    the path to the original binary file for which we're resolving
+  @crc        the expected checksum of the debuginfo file
+  @debug_file receives the path to the file, or NULL if not found
+	
+  @returns    1 on success, 0 if not found
+*/
+static int find_debug_file(const char *name, const char *binpath,
+                           unsigned int crc, char **debug_file) {
+  char fullpath[PATH_MAX];
+  char *bindir = NULL;
+
+  *debug_file = NULL;
+
+  bindir = strdup(binpath);
+  bindir = dirname(bindir);
+
+  // Search for the file in 'binpath'
+  sprintf(fullpath, "%s/%s", bindir, name);
+  if (access(fullpath, F_OK) != -1) {
+    *debug_file = strdup(fullpath);
+    goto DONE;
+  }
+
+  // Search for the file in 'binpath'/.debug
+  sprintf(fullpath, "%s/.debug/%s", bindir, name);
+  if (access(fullpath, F_OK) != -1) {
+    *debug_file = strdup(fullpath);
+    goto DONE;
+  }
+
+  // Search for the file in the global debug directory /usr/lib/debug/'binpath'
+  sprintf(fullpath, "/usr/lib/debug%s/%s", bindir, name);
+  if (access(fullpath, F_OK) != -1) {
+    *debug_file = strdup(fullpath);
+    goto DONE;
+  }
+
+DONE:
+  free(bindir);
+  return (*debug_file ? 1 : 0) && verify_checksum(*debug_file, crc);
+}
+
+static int find_debug_buildid(const char *buildid, char **debug_file ) {
+  char fullpath[PATH_MAX];
+
+  // Search for the file in the global debug directory with a sub-path:
+  // 	mm/nnnnnn...nnnn.debug
+  // Where mm are the first two characters of the buildid, and nnnn are the
+  // rest of the build id, followed by .debug.	
+  sprintf(fullpath, "/usr/lib/debug/.build-id/%c%c/%s.debug",
+          buildid[0], buildid[1],	buildid + 2);
+  if (access(fullpath, F_OK) != -1) {
+    *debug_file = strdup(fullpath);
+    return 1;
+  }
+
+  *debug_file = NULL;
+  return 0;
+}
+
+static int foreach_sym_core(const char *path, bcc_elf_symcb callback,
+                            void *payload, int is_debug_file) {
   Elf *e;
   int fd, res;
+  char *debug_name, *debug_file;
+  char buildid[128];	// currently 40 seems to be the default, let's be safe
+  unsigned int crc;
 
   if (openelf(path, &e, &fd) < 0)
     return -1;
 
+  // If there is a separate debuginfo file, try to locate and read it, first
+  // using the build-id section, then using the debuglink section. These are
+  // also the rules that GDB folows.
+  // See: https://sourceware.org/gdb/onlinedocs/gdb/Separate-Debug-Files.html
+  if (!is_debug_file && (
+      (
+       find_buildid(e, buildid) &&
+       find_debug_buildid(buildid, &debug_file))
+      ||
+      (
+       find_debuglink(e, &debug_name, &crc) &&
+       find_debug_file(debug_name, path, crc, &debug_file))
```

Comment:
would it be cleaner to do:
debug_file = find_debug_via_buildid(e);
if (!debug_file)
  debug_file = find_debug_file(e);